File: sdks/python/starfish/messaging.py
from __future__ import annotations

from typing import Any

from .connection import Connection
from .id import next_id
from .session import Session
from .types import DeliveryOptions, HeaderOptions, StarfishFrame, StarfishHeader


class Messaging:
    def __init__(self, connection: Connection, session: Session) -> None:
        self._connection = connection
        self._session = session

    async def send(
        self, to: str | list[str], payload: Any, options: HeaderOptions | None = None
    ) -> None:
        session_name = self._require_session()
        header = StarfishHeader(
            id=next_id("send"),
            resource="message",
            method="send",
            kind="request",
            session=session_name,
            to=to,
        )
        if options:
            header.delivery = options.delivery
            header.priority = options.priority
            header.ttl = options.ttl
            header.meta = options.meta

        frame = StarfishFrame(header=header, payload=payload)

        await self._connection.send(frame)

    async def broadcast(
        self, payload: Any, *, include_self: bool = False, options: HeaderOptions | None = None
    ) -> None:
        session_name = self._require_session()

        header = StarfishHeader(
            id=next_id("bcast"),
            resource="session",
            method="broadcast",
            kind="request",
            session=session_name,
        )

        if options:
            header.delivery = options.delivery
            header.priority = options.priority
            header.ttl = options.ttl
            header.meta = options.meta

        if include_self:
            if header.delivery is None:
                header.delivery = DeliveryOptions(include_self=True)
            else:
                header.delivery.include_self = True

        frame = StarfishFrame(header=header, payload=payload)

        await self._connection.send(frame)

    def _require_session(self) -> str:
        session = self._session.current
        if not session:
            raise RuntimeError("Not in a session. Call join() first.")
        return session

File: sdks/golang/starfish/messaging.go
package starfish

import (
	"context"
	"time"
)

// SendOptions configures a direct message or broadcast.
type SendOptions struct {
	Delivery *DeliveryOptions
	Priority string
	TTL      *int64
}

// messagingManager handles direct messaging and broadcast.
type messagingManager struct {
	conn    *connection
	idg     *IDGenerator
	session func() string
}

func newMessagingManager(conn *connection, idg *IDGenerator, session func() string) *messagingManager {
	return &messagingManager{
		conn:    conn,
		idg:     idg,
		session: session,
	}
}

// send delivers a direct message to one or more peers.
func (m *messagingManager) send(ctx context.Context, to string, payload map[string]any, opts *SendOptions) error {
	ts := time.Now().UnixMilli()
	frame := &Frame{
		Header: Header{
			V:        1,
			ID:       m.idg.Next("send"),
			Resource: "message",
			Method:   "send",
			Kind:     "request",
			Session:  m.session(),
			To:       SingleTarget(to),
			Ts:       &ts,
		},
		Payload: payload,
	}

	if opts != nil {
		frame.Header.Delivery = opts.Delivery
		frame.Header.Priority = opts.Priority
		frame.Header.TTL = opts.TTL
	}

	return m.conn.send(ctx, frame)
}

// sendMulti delivers a direct message to multiple peers.
func (m *messagingManager) sendMulti(ctx context.Context, to []string, payload map[string]any, opts *SendOptions) error {
	ts := time.Now().UnixMilli()
	frame := &Frame{
		Header: Header{
			V:        1,
			ID:       m.idg.Next("send"),
			Resource: "message",
			Method:   "send",
			Kind:     "request",
			Session:  m.session(),
			To:       MultiTarget(to...),
			Ts:       &ts,
		},
		Payload: payload,
	}

	if opts != nil {
		frame.Header.Delivery = opts.Delivery
		frame.Header.Priority = opts.Priority
		frame.Header.TTL = opts.TTL
	}

	return m.conn.send(ctx, frame)
}

// broadcast sends a message to all peers in the session.
func (m *messagingManager) broadcast(ctx context.Context, payload map[string]any, opts *SendOptions) error {
	ts := time.Now().UnixMilli()
	frame := &Frame{
		Header: Header{
			V:        1,
			ID:       m.idg.Next("broadcast"),
			Resource: "session",
			Method:   "broadcast",
			Kind:     "request",
			Session:  m.session(),
			Ts:       &ts,
		},
		Payload: payload,
	}

	if opts != nil {
		frame.Header.Delivery = opts.Delivery
		frame.Header.Priority = opts.Priority
		frame.Header.TTL = opts.TTL
	}

	return m.conn.send(ctx, frame)
}
